File: 2-rate-limiting-with-queues/2-csharp/ThrottledWorker.cs
using StackExchange.Redis;

namespace RateLimitApi;

/// <summary>
/// Worker-layer limiter: a leaky-bucket pull loop drains the shared Redis list
/// at most LimitMax jobs per window then paces — protecting downstream.
/// </summary>
public sealed class ThrottledWorker : BackgroundService
{
    public const string QueueKey = "q:tasks";
    public const string CompletedKey = "q:tasks:completed";

    private readonly IDatabase _redis;
    private readonly int _max;
    private readonly int _windowMs;

    public ThrottledWorker(IConnectionMultiplexer mux, IConfiguration config)
    {
        _redis = mux.GetDatabase();
        _max = config.GetValue("Worker:LimitMax", 10);
        _windowMs = config.GetValue("Worker:LimitDurationMs", 1000);
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        Console.WriteLine($"worker started: limiter {_max}/{_windowMs}ms");
        while (!stoppingToken.IsCancellationRequested)
        {
            var start = DateTimeOffset.UtcNow;
            for (var i = 0; i < _max; i++)
            {
                var job = await _redis.ListLeftPopAsync(QueueKey);
                if (job.IsNull)
                {
                    break;
                }
                await _redis.StringIncrementAsync(CompletedKey);
                Console.WriteLine($"job {job} completed at {DateTimeOffset.UtcNow:O}");
            }
            var elapsed = (DateTimeOffset.UtcNow - start).TotalMilliseconds;
            if (elapsed < _windowMs)
            {
                await Task.Delay(_windowMs - (int)elapsed, stoppingToken);
            }
        }
    }
}

File: 2-rate-limiting-with-queues/1-java/src/main/java/com/starci/ratelimit/ThrottledWorker.java
package com.starci.ratelimit;

import java.time.Instant;

import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Component;

/**
 * Worker-layer limiter: a leaky-bucket pull loop drains the shared Redis list
 * at most limitMax jobs per window, then paces — protecting downstream.
 * Only active when app.role=worker.
 */
@Component
public class ThrottledWorker {

    private final StringRedisTemplate redis;
    private final AppProperties props;

    public ThrottledWorker(StringRedisTemplate redis, AppProperties props) {
        this.redis = redis;
        this.props = props;
    }

    @EventListener(ApplicationReadyEvent.class)
    public void start() {
        if (!"worker".equals(props.getRole())) {
            return;
        }
        int max = props.getWorker().getLimitMax();
        long window = props.getWorker().getLimitDurationMs();
        Thread t = new Thread(() -> runLoop(max, window), "throttled-worker");
        t.setDaemon(true);
        t.start();
        System.out.printf("worker started: limiter %d/%dms%n", max, window);
    }

    private void runLoop(int max, long window) {
        while (true) {
            long start = System.currentTimeMillis();
            for (int i = 0; i < max; i++) {
                String job = redis.opsForList().leftPop(ApiController.QUEUE_KEY);
                if (job == null) {
                    break;
                }
                redis.opsForValue().increment(ApiController.COMPLETED_KEY);
                System.out.printf("job %s completed at %s%n", job, Instant.now());
            }
            long elapsed = System.currentTimeMillis() - start;
            if (elapsed < window) {
                try {
                    Thread.sleep(window - elapsed);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
            }
        }
    }
}
